driver/src/main/java/org/dcache/SapphireDriver.java

```diff
@@ -168,11 +366,52 @@ public void configure(Map<String, String> properties) throws IllegalArgumentExce
         }
         MongoDatabase mongoDatabase = mongoClient.getDatabase(database);
         files = mongoDatabase.getCollection("files");
+        stageFiles = mongoDatabase.getCollection("stage");
 
         long schedulerPeriod = Long.parseLong(properties.getOrDefault("period", "1"));
         TimeUnit periodUnit = TimeUnit.valueOf(properties.getOrDefault("period_unit", TimeUnit.MINUTES.name()));
 
         executorService.scheduleAtFixedRate(new FireAndForgetTask(this::processFlush), schedulerPeriod, schedulerPeriod, periodUnit);
+        executorService.scheduleAtFixedRate(new FireAndForgetTask(this::processStage), schedulerPeriod, schedulerPeriod, periodUnit);
+
+        String[] whitelist = properties.getOrDefault("whitelist", "").split(",");
+        String portStr = properties.getOrDefault("port", "");
+
+        int port;
+        if(portStr.contains("-")) {
+            _log.debug("Port range is used");
+            port = getFreePort(Integer.parseInt(portStr.split("-")[0]),
+                    Integer.parseInt(portStr.split("-")[1]));
+        } else {
+            _log.debug("Static port is used");
+            port = Integer.parseInt(portStr);
+        }
+        _log.info("Sapphire is running on port {}", port);
+
+        try {
+            server = new FileServer(port, whitelist);
+            server.startServer();
+        } catch (Exception e) {
+            _log.error("Could not start Jetty server", e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    private int getFreePort(int start, int end) {
+        _log.debug("Trying to find free port between {} and {}", start, end);
+        int port = start;
+        while (port < end) {
+            _log.debug("Trying port {}", port);
+            try {
+                new ServerSocket(port).close();
```

Comment:
after `close` another process might allocate  the selected port and binding to the socket will fail. Do we need port rage here at all?

Reply:
I needed to add support for port range because, at least in my test system, the plugin gets created twice